bluesky_browser/suitcase.py

```diff
@@ -0,0 +1,298 @@
+import traitlets
+from PyQt5.QtWidgets import (QPushButton, QVBoxLayout, QTabWidget, QCheckBox,
+                             QFormLayout, QLineEdit, QSpinBox)
+import suitcase.csv
+import suitcase.json_metadata
+import suitcase.specfile
+import suitcase.tiff_stack
+import suitcase.tiff_series
+
+from .utils import (ConfigurableQWidget, QTextListEdit, QIntListEdit,
+                    QFloatListEdit, QBoolBox, QBoolNoneBox)
+
+# A global dict is defined here to map QtWidgets to their value attribute name
+# this is required to construct the slot and signal attributes when these are
+# used.
+_value_map = {QLineEdit: 'text',
+              QTextListEdit: 'text',
+              QIntListEdit: 'text',
+              QFloatListEdit: 'text',
+              QBoolBox: 'currentText',
+              QBoolNoneBox: 'currentText',
+              QSpinBox: 'value'}
+
+
+class SuitcaseWidget(ConfigurableQWidget):
+    '''Widget for entering generic ``suitcase.xyz.export()`` parameters.
+
+    Parameters
+    ----------
+    *args : args
+        args passed down to QWidget
+    geometry : tuple
+        Tuple giving the (left, top, width, height) required for the window
+    **kwargs : kwargs
+        kwargs passed down to QWidget
+
+    Configurable Traits
+    -------------------
+    directory : str
+        The directory passed to the ``suitcase.xyz`` exporter
+    file_prefix : str
+        The file_prefix passed to the ``suitcase.xyz`` exporter
+    show : list
+        A list of parameters to include in the dialog box
+    '''
+
+    file_prefix = traitlets.Unicode('test_data_{start[scan_id]}')
+    file_prefix.tag(widget=QLineEdit)
+    directory = traitlets.Unicode('test_data')
+    directory.tag(widget=QLineEdit)
+    show = traitlets.List(['file_prefix', 'directory'])
+
+    def __init__(self, *args, geometry=(10, 10, 400, 140), **kwargs):
+        super().__init__(*args, **kwargs)
+        self.geometry = geometry
+        self.initUI()
+
+    def initUI(self):
+        self.setGeometry(*self.geometry)
+
+        self.main_layout = QFormLayout()
+
+        self.enable_checkbox = QCheckBox('enable exporter', self)
+        self.main_layout.addRow('', self.enable_checkbox)
+
+        # step thorugh each require input parameter
+        for parameter in self.show:
+            # collect the trait and widget_type objects.
+            trait = self.traits()[parameter]
+            widget_type = trait.metadata['widget']
+            # create, and reference, the widget.
+            setattr(self, f'{parameter}_widget', widget_type())
```

Comment:
A minor thing: this monkey-patching and string-munging stands out to me. Could we instead define `self.widgets = {}` in `__init__` and then use `widget = widget_type(); self.widgets[parameter] = widget` here?